src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -1192,137 +1243,180 @@ static void ProcessNode(RegexNode node, RegexNode subsequent, int maxDepth)
+        /// </summary>
+        /// <returns>The found node that should be explored further for auto-atomicity; null if it doesn't exist.</returns>
+        private static RegexNode? FindLastExpressionInLoopForAutoAtomic(RegexNode node, int maxDepth)
+        {
+            Debug.Assert(node.Type == Loop);
 
-                        // Skip the successor down to the guaranteed next node.
-                        while (subsequent.ChildCount() > 0)
-                        {
-                            Debug.Assert(subsequent.Type != Group);
-                            switch (subsequent.Type)
-                            {
-                                case Concatenate:
-                                case Capture:
-                                case Atomic:
-                                case Require:
-                                case Loop when subsequent.M > 0:
-                                case Lazyloop when subsequent.M > 0:
-                                    subsequent = subsequent.Child(0);
-                                    continue;
-                            }
+            // Start by looking at the loop's sole child.
+            node = node.Child(0);
 
-                            break;
-                        }
+            // Skip past captures.
+            while (node.Type == Capture)
+            {
+                node = node.Child(0);
+            }
 
-                        // If the two nodes don't agree on case-insensitivity, don't try to optimize.
-                        // If they're both case sensitive or both case insensitive, then their tokens
-                        // will be comparable.
-                        if ((node.Options & RegexOptions.IgnoreCase) != (subsequent.Options & RegexOptions.IgnoreCase))
-                        {
-                            return false;
-                        }
+            // If the loop's body terminates with a {one/notone/set} loop, return it.
+            if (node.Type == Oneloop || node.Type == Notoneloop || node.Type == Setloop)
+            {
+                return node;
+            }
 
-                        // If the successor is an alternation, all of its children need to be evaluated, since any of them
-                        // could come after this node.  If any of them fail the optimization, then the whole node fails.
-                        if (subsequent.Type == Alternate)
-                        {
-                            int childCount = subsequent.ChildCount();
-                            for (int i = 0; i < childCount; i++)
-                            {
-                                if (!CanBeMadeAtomic(node, subsequent.Child(i), maxDepth - 1))
-                                {
-                                    return false;
-                                }
-                            }
+            // If the loop's body is a concatenate, we can skip to its last child iff that
```

Comment:
For comments like this I'm mentally visualizing patterns, you might consider sprinkling more example patterns in comments.

Reply:
I've gone through and added a lot of such comments.